### ISSUE TITLE
14.4 Extern alias directives

### ISSUE THREAD

**Issue by RexJaeschke** [dotnet/csharpstandard#193]
As well as looking at the diffs between the md from both, it will be useful to also look at the (few) tracked changes TG2 made in this subclause (§14.4) of the WD3.17 Word document.

**V5std has**

“Within C# source code, a type is declared a member of a single namespace. However, a namespace hierarchy referenced by an extern alias may contain types that are also members of other namespaces. For example, if `A` and `B` are extern aliases, the names `A::X`,`B::C.Y` and `global::D.Z` may, depending on the external specification supported by the particular compiler, all refer to the same type.

The alias introduced by an *extern_alias_directive* is very similar to the alias introduced by a *using_alias_directive*. See [§14.5.2](#using-alias-directives) for more detailed discussion of *extern_alias_directive*s and *using_alias_directive*s.

`alias` is a contextual keyword ([§7.4.4](lexical-structure.md#keywords)) and only has special meaning when it immediately follows the `extern` keyword in an *extern_alias_directive*.

> [*Example*: In fact an extern alias could use the identifier `alias` as its name:
> ```csharp
> extern alias alias;
> ```
> *end example*]”

**While v6spec has**

“An *extern_alias_directive* makes an alias available within a particular compilation unit or namespace body, but it does not contribute any new members to the underlying declaration space. In other words, an *extern_alias_directive* is not transitive, but, rather, affects only the compilation unit or namespace body in which it occurs.

The following program declares and uses two extern aliases, `X` and `Y`, each of which represent the root of a distinct namespace hierarchy:
```csharp
extern alias X;
extern alias Y;
class Test
{
    X::N.A a;
    X::N.B b1;
    Y::N.B b2;
    Y::N.C c;
}
```

The program declares the existence of the extern aliases `X` and `Y`, but the actual definitions of the aliases are external to the program. The identically named `N.B` classes can now be referenced as `X.N.B` and `Y.N.B`, or, using the namespace alias qualifier, `X::N.B` and `Y::N.B`. An error occurs if a program declares an extern alias for which no external definition is provided.“

Which is correct?

**Comment by Nigel-Ecma** [comment]
The v5 & v6 wordings are mentioning different aspects of extern aliases:

- The v5 version is stating that the same type might end up accessible by different names. Note that it does give any implications of this.
- The v6 version is talking about extern aliases not being transitive, which I understand to mean that if namespace introduced via an extern alias `A` itself contains an extern alias `B`, then the latter is `A.B` and not introduced into the current declaration space as simply `B`. Note that having made this point the following example doesn't demonstrate it.

If the above is correct then *both* v5 & v6 are "correct", both also seem a bit lacking (the two notes), and maybe we should merge v5 & v6 and, if needed, address the points in the above notes.

**Comment by Nigel-Ecma** [comment]
From @Nigel-Ecma Mar 10 above:

> * The v6 version is talking about extern aliases not being transitive, which I understand to mean that if namespace introduced via an extern alias `A` itself contains an extern alias `B`, then the latter is `A.B` and not introduced into the current declaration space as simply `B`. Note that having made this point the following example doesn't demonstrate it.

Re-reading the v6 text I'm not sure it is saying `B` is available at all even as `A.B`.

In contrast if we consider the `#include/import` type statements of C-like languages then the operation *is* transitive – import a file which itself contains imports then those too are imported, etc. The v6 text is stating this is not the case in C#. However transitivity is there for a good reason in those other languages. Either this issue is dealt with elsewhere in the Standard and the v6 text shouldn't include it here, or the v6 text needs clarification, or...?

**Comment by jskeet** [comment]
> Re-reading the v6 text I'm not sure it is saying B is available at all even as A.B.

Indeed. And I think that's okay. An extern alias is effectively a private implementation detail, usually used to get around otherwise-awkward duplication (such as the same namespace-qualified type name being present in two assemblies).

I'm personally okay with this part, and think the v6 spec is clear enough on it. However, what's not clear to me is whether the example in the v5 spec should also be reintroduced - I'm not sure that it's ever actually been valid. I think we need further discussion on this.

**Comment by jskeet** [comment]
Note: I'll leave meeting-discuss here as we really *should* discuss this, but as the conversation is currently primarily between myself and Nigel, and Nigel won't be at the meeting, I suggest we punt on this one for today.

**Comment by jskeet** [comment]
Revisiting this now... it looks like we still have the v5 text in 13.4, but none of the C# 6 text.

It's not even clear to me that 13.4 starts correctly:

> An extern_alias_directive introduces an identifier that serves as an alias for a namespace.

Is that actually the case? I've only ever seen extern aliases effectively used to refer to the "root" of another assembly.

Indeed, the [external alias reference page](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/extern-alias) refers to it as an "external assembly alias".

Is there any way to refer to a specific *namespace* within another assembly?

The whole of 13.4 seems questionable to me at the moment.

I've lost track of where the original draft C# 6 spec was, but I would suggest we take that as a starting point, and see what we think it's actually missing.

Rex, can you create a PR along those lines, and we can discuss it in the meeting?

(Assigning to Rex.)

**Comment by RexJaeschke** [comment]
@jskeet This is well beyond me, so I am dropping the assignment. The most capable person to write this up would seem to be you or @Nigel-Ecma.

**Comment by jskeet** [comment]
@RexJaeschke: Sorry, to clarify - in the very first comment, you seem to have a draft C# 6 spec which you were comparing with what we've got. I can't find that now, and the history across multiple repos is going to be very hard to follow; I'm hoping that you've still got it somewhere and could basically create a PR that replaces our existing 13.4 with the one that was in the draft C# 6 spec from May 2020.

**Comment by Nigel-Ecma** [comment]
I’ll postulate what has happened here, Mads can correct me!

The original description says:

The specification of the aliased namespace is external to the source code of the program and applies also to nested namespaces of the aliased namespace.

and I postulate that the VS implementation of this external spec was to introduce a namespace for the whole of an assembly, and not to support any other mapping. That fits the Standard’s description, so everything is conforming.

VS’s external spec was then fed back into the MS C# Reference and the feature became to be described as an assembly alias as that all VS supported.

So the Standard isn’t wrong, VS is conforming, but the description doesn’t really match the experience most/all C# users get and so it seems to be wrong…

I’ll take a wild guess that no other C# implementation exists that has done anything other than match the VS implementation, so the Standard could be rewritten to match that (removing possibilities that no one has implemented).

But if it is to go in the Standard what do we call it? The concept that an assembly can be treated as a namespace doesn’t currently exist does it?

**Comment by jskeet** [comment]
> So the Standard isn’t wrong, VS is conforming, but the description doesn’t really match the experience most/all C# users get and so it seems to be wrong…

It's slightly weirder than that... while the common approach used to be adding references to other assemblies, it's now to add package references (i.e. to NuGet packages). I can't find anything in Visual Studio that lets me specify an extern alias through the GUI.

It turns out to be pretty simple to do in the csproj file, once you've found the right blog post. Here's a sample:

Project.csproj:
```xml
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net6.0</TargetFramework>
  </PropertyGroup>

  <ItemGroup>
    <PackageReference Include="NodaTime" Version="3.0.9" Aliases="noda"/>
  </ItemGroup>
</Project>
```

LocalDate.cs:

```csharp
namespace NodaTime
{
    public class LocalDate
    {
        public void NotInNodaTime() { }
    }
}
```

And Program.cs:

```csharp
extern alias noda;
using noda::NodaTime;

LocalDate date = new LocalDate();
System.Console.WriteLine(date.AtMidnight());
```

But again, this seems to be for the whole assembly (or possibly every assembly in the package - not sure what happens if there are multiple assemblies). I don't think there's any way of saying that the extern alias is "The NodaTime.Text namespace within the NodaTime.dll assembly" for example.

**Comment by Nigel-Ecma** [comment]
Which suggests that the original scribe had foresight (I hasten to add it wasn’t me! ;-)) to write that the mechanism is "external to the source code of the program”, i.e. not the business of the Standard.

First we have VS providing a mechanism to alias an assembly as a namespace. Now, if you edit csproj files, you can alias a nuget package. Next…?

But none of this is “C#” per se, its part of the compilation system. From a C# perspective a namespace is introduced; whether it directly maps to one, or to a “virtual” one representing the whole of an assembly or nuget package doesn’t change this.

I think the question is do we amend the Standard text in some way to better reflect how this is currently implemented – example, implementation specific note?

The answer maybe yes, but in v7 maybe?

Hopefully any proposed change will not need to mention editing csproj files!

**Comment by RexJaeschke** [comment]
It seems that there is still confusion about how to resolve this. In any event, below is the verbatim text for this section from the MS V6 spec:

## Extern aliases

An *extern_alias_directive* introduces an identifier that serves as an alias for a namespace. The specification of the aliased namespace is external to the source code of the program and applies also to nested namespaces of the aliased namespace.

```antlr
extern_alias_directive
    : 'extern' 'alias' identifier ';'
    ;
```

The scope of an *extern_alias_directive* extends over the *using_directive*s, *global_attributes* and *namespace_member_declaration*s of its immediately containing compilation unit or namespace body.

Within a compilation unit or namespace body that contains an *extern_alias_directive*, the identifier introduced by the *extern_alias_directive* can be used to reference the aliased namespace. It is a compile-time error for the *identifier* to be the word `global`.

An *extern_alias_directive* makes an alias available within a particular compilation unit or namespace body, but it does not contribute any new members to the underlying declaration space. In other words, an *extern_alias_directive* is not transitive, but, rather, affects only the compilation unit or namespace body in which it occurs.

The following program declares and uses two extern aliases, `X` and `Y`, each of which represent the root of a distinct namespace hierarchy:
```csharp
extern alias X;
extern alias Y;

class Test
{
    X::N.A a;
    X::N.B b1;
    Y::N.B b2;
    Y::N.C c;
}
```

The program declares the existence of the extern aliases `X` and `Y`, but the actual definitions of the aliases are external to the program. The identically named `N.B` classes can now be referenced as `X.N.B` and `Y.N.B`, or, using the namespace alias qualifier, `X::N.B` and `Y::N.B`. An error occurs if a program declares an extern alias for which no external definition is provided.

**Comment by Nigel-Ecma** [comment]
This starts the same, up to the para after the grammar, as what we currently have in the Draft. The example at the end is much better than the:

For example, if A and B are extern aliases, the names A::X,B::C.Y and global::D.Z may, depending on the external specification supported by the particular compiler, all refer to the same type.

in the Draft.

IIRC there was a discussion over the transitive bit below, offhand I can’t remember the conclusion, will go digging…

I think we could merge these two versions for v6 without too much difficulty and the better example would be an improvement.

**Comment by Nigel-Ecma** [comment]
OK, didn't need to dig far now I'm logged into GitHub:

It was me a year ago who raised the transitive bit, and also suggested we merge the two versions back then – reassuring (to me) :-)

**Comment by Nigel-Ecma** [comment]
First past at a merge, just change the last example from `extern alias alias` to the more useful one:

---

## 13.4 Extern alias directives

An *extern_alias_directive* introduces an identifier that serves as an alias for a namespace. The specification of the aliased namespace is external to the source code of the program and applies also to nested namespaces of the aliased namespace.

```ANTLR
extern_alias_directive
    : 'extern' 'alias' identifier ';'
    ;
```

The scope of an *extern_alias_directive* extends over the *using_directive*s, *global_attributes* and *namespace_member_declaration*s of its immediately containing *compilation_unit* or *namespace_body*.

Within a compilation unit or namespace body that contains an *extern_alias_directive*, the identifier introduced by the *extern_alias_directive* can be used to reference the aliased namespace. It is a compile-time error for the *identifier* to be the word `global`.

Within C# source code, a type is declared a member of a single namespace. However, a namespace hierarchy referenced by an extern alias may contain types that are also members of other namespaces. For example, if `A` and `B` are extern aliases, the names `A::X`,`B::C.Y` and `global::D.Z` may, depending on the external specification supported by the particular compiler, all refer to the same type.

The alias introduced by an *extern_alias_directive* is very similar to the alias introduced by a *using_alias_directive*. See [§13.5.2](namespaces.md#1352-using-alias-directives) for more detailed discussion of *extern_alias_directive*s and *using_alias_directive*s.

`alias` is a contextual keyword ([§6.4.4](lexical-structure.md#644-keywords)) and only has special meaning when it immediately follows the `extern` keyword in an *extern_alias_directive*.

> *Example*: The following program declares and uses two extern aliases, `X` and `Y`, each of which represent the root of a distinct namespace hierarchy:

> ```csharp
> extern alias X;
> extern alias Y;
> 
> class Test
> {
>     X::N.A a;
>     X::N.B b1;
>     Y::N.B b2;
>     Y::N.C c;
> }
> ```

> The program declares the existence of the extern aliases `X` and `Y`, but the actual definitions of the aliases are external to the program. The identically named `N.B` classes can now be referenced as `X.N.B` and `Y.N.B`, or, using the namespace alias qualifier, `X::N.B` and `Y::N.B`. An error occurs if a program declares an extern alias for which no external definition is provided.
> *end example*

---

- The transitive bit is already in the reference §13.5.2
- I've left in the confusing “`A::X`,`B::C.Y` and `global::D.Z`” bit, it is making the point that different names can refer to the same thing but I think it isn't doing it well. Could do with a better explained example and probably a note on any consequences of this?
- None of this addresses the “assembly/package presented as a namespace” issue that Jon raised.

**Comment by jskeet** [comment]
Suggested change to the example:

`A::Y.Z`, `B::Z`, `global::X.Y.Z`

I can see how *that* could work... if `A::X.Z` and `B::Y.Z` can both refer to the same type that's slightly trickier, but we *could* potentially use it.

**Comment by jskeet** [comment]
This bit should be normative, before the example: 

> An error occurs if a program declares an extern alias for which no external definition is provided.

As we don't understand the example containing `D.Z`, and the paragraph containing it appears to try to avoid people assuming something that would be odd to assume anyway, let's remove that paragraph.

**Comment by jskeet** [comment]
We discussed whether the compiler needs to be aware of type identity. It appears to be:

```csharp
// In the project file: <PackageReference Include="NodaTime" Version="3.0.9" Aliases="noda,noda2"/>
extern alias noda;
extern alias noda2;

noda::NodaTime.LocalDate date = new noda2::NodaTime.LocalDate();
```
